src/jvm/main/org/jetbrains/kotlinx/lincheck/transformation/LincheckJavaAgent.kt

```diff
@@ -310,27 +329,39 @@ internal object LincheckClassFileTransformer : ClassFileTransformer {
         }
 
     override fun transform(
-        loader: ClassLoader?, className: String, classBeingRedefined: Class<*>?, protectionDomain: ProtectionDomain?, classBytes: ByteArray
+        loader: ClassLoader?,
+        internalClassName: String,
+        classBeingRedefined: Class<*>?,
+        protectionDomain: ProtectionDomain?,
+        classBytes: ByteArray
     ): ByteArray? = runInIgnoredSection {
-        if (instrumentationMode == MODEL_CHECKING && !INSTRUMENT_ALL_CLASSES_IN_MODEL_CHECKING_MODE) {
-            // In the model checking mode, we transform classes lazily,
-            // once they are used in the testing code.
-            if (className.canonicalClassName !in instrumentedClassesInTheModelCheckingMode) return null
-        } else {
-            if (!shouldTransform(className.canonicalClassName, instrumentationMode)) return null
+        // If the class should not be transformed, return immediately.
+        if (!shouldTransform(internalClassName.canonicalClassName, instrumentationMode)) {
+            return null
+        }
+        // In the model checking mode, we transform classes lazily,
+        // once they are used in the testing code.
+        if (!INSTRUMENT_ALL_CLASSES &&
+            instrumentationMode == MODEL_CHECKING &&
+            internalClassName.canonicalClassName !in instrumentedClasses) {
+            return null
         }
-        return transformImpl(loader, className, classBytes)
+        return transformImpl(loader, internalClassName, classBytes)
     }
 
-    private fun transformImpl(loader: ClassLoader?, className: String, classBytes: ByteArray): ByteArray = transformedClassesCache.computeIfAbsent(className) {
-        nonTransformedClasses[className] = classBytes
+    private fun transformImpl(
+        loader: ClassLoader?,
+        internalClassName: String,
+        classBytes: ByteArray
+    ): ByteArray = transformedClassesCache.computeIfAbsent(internalClassName.canonicalClassName) {
+        nonTransformedClasses.putIfAbsent(internalClassName.canonicalClassName, classBytes.copyOf())
```

Comment:
Should we need to call `copyOf()`? If so, please explain why in the comment.

Reply:
I removed it. 

I think I've added it during debugging, but it is actually not needed.